core/iwasm/common/wasm_runtime_common.c

```diff
@@ -194,7 +194,6 @@ runtime_signal_handler(void *sig_addr)
         else if (exec_env_tls->exce_check_guard_page <= (uint8 *)sig_addr
                  && (uint8 *)sig_addr
                         < exec_env_tls->exce_check_guard_page + page_size) {
-            bh_assert(wasm_get_exception(module_inst));
```

Comment:
Had better not remove and modify to:
```C
    bh_assert(wasm_get_exception(module_inst)
#if WASM_ENABLE_THREAD_MGR != 0
              || wasm_cluster_is_thread_terminated(exec_env_tls));
#endif
);
```
And also change the similar place for Windows (this file, L253):
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/common/wasm_runtime_common.c#L253

Reply:
Right, I'm debugging to understand why `exec_env_tls->suspend_flags` is not set properly once I get there.

Reply:
Ok, the problem was that we don't set the suspend flag for the thread that originally raises the exception.

Reply:
The problem is that the assert would be executed also on the thread that originally raises the exception and that thread doesn't have the suspend flag set. I tried setting the suspend flag on that initial thread too but it break things since it forces the initial thread to exit prematurely (after a `CHECK_SUSPEND_FLAGS();`) while we don't want that.

So I was thinking of just having a
```c
#if WASM_ENABLE_THREAD_MGR == 0
    bh_assert(wasm_get_exception(module_inst));
#endif
```

Reply:
Not sure how to reproduce the issue? The thread_terminate sample with TEST_TERMINATION_BY_TRAP == 1 and TEST_TERMINATION_IN_MAIN_THREAD == 1?

Reply:
I updated the thread termination sample, you should be able to reproduce the problem (in AOT mode) by running it (no changes required). Let me know if you have issues reproducing it.

Reply:
Thanks @eloparco, I can reproduce the issue now. The process is that:
the main thread met wasi_proc_exit, raised that exception, and spread it to child threads
the main thread accessed the guard page to trigger signal handler, but hadn't entered the signal handler
the child thread got exception and ended running, and it founds that the exception is `wasi proc exit` excpeiton, so it tried to clear it, and during the clearing, **it spread null exception to other threads including main thread**
the main thread run into signal handler, but the exception had been cleared, so bh_assert failed

So the root cause is that the main thread's `wasi proc exit` exception is cleared by child thread.
It seems a better solution is not to spread the exception `wasi proc exit` to other threads, but just set the suspend flags of other threads to let them exit.

I think we may merge this PR firstly if you are OK and we will submit another PR to fix that issue.

Reply:
> I think we may merge this PR firstly if you are OK and we will submit another PR to fix that issue.

Let's do it, so that I can move to #1985 that hopefully should succeed after we merge this.

Reply:
@eloparco I tried again, and think this is not a best way to resolve the issue. The root cause is that the main thread's "wasi proc exit" exception is cleared by the sub thread after the latter gets the spread exception from main thread and then clear exception for all threads.
It is strange to spread the "wasi proc exit" exception to other threads and then clear it in other threads. I think we can just ignore spreading of this exception and no need to clear it again. I uploaded PR #1988 to fix the issues, I tested the cases and didn't find the issues you mentioned (unreachable exception was thrown and bh_assert failure), could you help review and try? Thanks.

Reply:
Can we merge this one and rebase #1988? I'll then try #1988 in a few different cases to see if we don't have regressions.

Reply:
My personal preference is we had better not merge a temporarily PR which doesn't fix the root cause, #1988 should be better and have resolved the issues mentioned in this PR. I tend to merge #1988 directly.

Reply:
Right, it makes sense. Let's close this one once #1988 gets merged